File: src/main/java/com/freya02/botcommands/api/localization/context/AppLocalizationContext.java
package com.freya02.botcommands.api.localization.context;

import com.freya02.botcommands.api.localization.Localization;
import net.dv8tion.jda.api.interactions.DiscordLocale;
import net.dv8tion.jda.api.interactions.Interaction;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.annotation.CheckReturnValue;

/**
 * Interface helping in localizing content, supports preset localization bundles,
 * localization prefixes and context-aware localization.
 *
 * <p>This interface also includes the user's and the guild's locale by default.
 *
 * <p>While instances of this interface are primarily injected,
 * you can also construct instances of this interface with {@link #create(String, String, DiscordLocale, DiscordLocale)}.
 *
 * @see #getUserLocale()
 * @see #getGuildLocale()
 *
 * @see #create(String, String, DiscordLocale, DiscordLocale)
 */
public interface AppLocalizationContext extends TextLocalizationContext {
    /**
     * Returns the Locale of the user
     * <br>The locale can either come from the {@link Interaction} or from {@link LocalizationContext#withGuildLocale(DiscordLocale)}.
     *
     * @return The Locale of the user
     *
     * @see #withUserLocale(DiscordLocale)
     */
    @NotNull //User locale is always provided in interactions
    DiscordLocale getUserLocale();

    @NotNull
    @Override
    @CheckReturnValue
    AppLocalizationContext withGuildLocale(@Nullable DiscordLocale guildLocale);

    @NotNull
    @Override
    @CheckReturnValue
    AppLocalizationContext withBundle(@NotNull String localizationBundle);

    @NotNull
    @Override
    @CheckReturnValue
    AppLocalizationContext withPrefix(@Nullable String localizationPrefix);

    /**
     * Localizes the provided path, in the specified bundle, with the user's locale
     *
     * @param localizationBundle The name of the localization bundle
     * @param localizationPath   The localization path to search for
     * @param entries            The entries to fill the template
     *
     * @return The localized string
     */
    @NotNull
    default String localizeUser(@NotNull String localizationBundle, @NotNull String localizationPath, @NotNull Localization.Entry @NotNull ... entries) {
        return localize(getUserLocale(), localizationBundle, localizationPath, entries);
    }

    /**
     * Localizes the provided path, in the current context's bundle, with the user's locale
     *
     * @param localizationPath The localization path to search for
     * @param entries          The entries to fill the template
     *
     * @return The localized string
     */
    @NotNull
    default String localizeUser(@NotNull String localizationPath, @NotNull Localization.Entry @NotNull ... entries) {
        return localize(getUserLocale(), getLocalizationBundle(), localizationPath, entries);
    }
}

File: src/main/kotlin/com/freya02/botcommands/api/localization/context/AppLocalizationContext.kt
package com.freya02.botcommands.api.localization.context

import com.freya02.botcommands.api.localization.Localization
import com.freya02.botcommands.api.localization.annotations.LocalizationBundle
import net.dv8tion.jda.api.interactions.DiscordLocale
import net.dv8tion.jda.api.interactions.Interaction
import javax.annotation.CheckReturnValue

/**
 * Interface helping in localizing content, supports preset localization bundles,
 * localization prefixes, and context-aware localization.
 *
 * This interface also includes the user's and the guild's locale by default.
 *
 * While instances of this interface are primarily injected with [LocalizationBundle],
 * you can also construct instances of this interface with [LocalizationContext.create].
 *
 * Instances are only injectable if the event is a subclass of either [Interaction].
 *
 * @see userLocale
 * @see guildLocale
 * @see LocalizationContext.create
 */
interface AppLocalizationContext : TextLocalizationContext {
    /**
     * Returns the Locale of the user.
     *
     * The locale can either come from the [Interaction] or from [LocalizationContext.withGuildLocale].
     *
     * @return The Locale of the user
     *
     * @see withUserLocale
     */
    //User locale is always provided in interactions
    val userLocale: DiscordLocale

    @CheckReturnValue
    override fun withGuildLocale(guildLocale: DiscordLocale?): AppLocalizationContext

    @CheckReturnValue
    override fun withBundle(localizationBundle: String): AppLocalizationContext

    @CheckReturnValue
    override fun withPrefix(localizationPrefix: String?): AppLocalizationContext

    @CheckReturnValue
    override fun switchBundle(localizationBundle: String): AppLocalizationContext

    /**
     * Localizes the provided path, with the user's locale.
     *
     * @param localizationPath The path of the localization template, prefixed with [localizationPrefix][LocalizationContext.localizationPrefix]
     * @param entries          The entries to fill the template with
     *
     * @return The localized string
     */
    fun localizeUser(localizationPath: String, vararg entries: Localization.Entry): String =
        localize(userLocale, localizationPath, *entries)

    /**
     * Localizes the provided path, with the user's locale, or returns `null` if the path does not exist.
     *
     * @param localizationPath The path of the localization template, prefixed with [localizationPrefix][LocalizationContext.localizationPrefix]
     * @param entries          The entries to fill the template with
     *
     * @return The localized string
     */
    fun localizeUserOrNull(localizationPath: String, vararg entries: Localization.Entry): String? =
        localizeOrNull(userLocale, localizationPath, *entries)
}

/**
 * Localizes the provided path, with the user's locale.
 *
 * @param localizationPath The path of the localization template, prefixed with [localizationPrefix][LocalizationContext.localizationPrefix]
 * @param entries          The entries to fill the template with
 *
 * @return The localized string
 */
fun AppLocalizationContext.localizeUser(localizationPath: String, vararg entries: PairEntry): String =
    localize(userLocale, localizationPath, *entries.mapToEntries())

/**
 * Localizes the provided path, with the user's locale, or returns `null` if the path does not exist.
 *
 * @param localizationPath The path of the localization template, prefixed with [localizationPrefix][LocalizationContext.localizationPrefix]
 * @param entries          The entries to fill the template with
 *
 * @return The localized string
 */
fun AppLocalizationContext.localizeUserOrNull(localizationPath: String, vararg entries: PairEntry): String? =
    localizeOrNull(userLocale, localizationPath, *entries.mapToEntries())
